data/transactions/transaction.go

```diff
@@ -263,6 +299,66 @@ func (tx Transaction) WellFormed(spec SpecialAddresses, proto config.ConsensusPa
 		if !proto.Asset {
 			return fmt.Errorf("asset transaction not supported")
 		}
+	case protocol.ApplicationCallTx:
+		if !proto.Application {
+			return fmt.Errorf("application transaction not supported")
+		}
+
+		// Ensure requested action is valid
+		switch tx.OnCompletion {
+		case NoOpOC:
+		case OptInOC:
+		case CloseOutOC:
+		case ClearStateOC:
+		case UpdateApplicationOC:
+		case DeleteApplicationOC:
+		default:
+			return fmt.Errorf("invalid application OnCompletion")
+		}
+
+		if tx.ApplicationID != 0 && tx.OnCompletion != UpdateApplicationOC {
+			// Ensure programs are only set for creation or update
+			if tx.ApprovalProgram != "" || tx.ClearStateProgram != "" {
+				return fmt.Errorf("scripts may only be specified during application creation or update")
+			}
+		}
+
+		if tx.ApplicationID != 0 {
+			if tx.LocalStateSchema != (basics.StateSchema{}) ||
```

Comment:
I think we should compare with an empty struct like this wherever possible, because it's likely the compiler can optimize it well. I only added `Empty` methods when they are necessary because the struct contains a reference type.